File: kits/python/src/ironmcp/results.py
"""Content/result helpers — the generic PIPING an app tool rides on.

The screenshot/audio CAPTURE stays app-side; ironmcp owns how raw bytes become a
well-formed, guarded MCP result. No helper echoes a caller-supplied value.

A WSLg/X11 capture can exit 0 yet emit an empty (``<=8``-byte) file; every binary
helper treats that as a FAILURE, not media (loqu8 invariant #8). Ported from the Dart
kit's ``Results`` (``kits/dart/lib/src/results.dart``) — same contract, Python idiom.
"""

from __future__ import annotations

import base64
import json as _json

from mcp.types import AudioContent, CallToolResult, ImageContent, TextContent

__all__ = ["Results", "MIN_BYTES"]

# Minimum bytes that count as real payload (loqu8 invariant #8).
MIN_BYTES = 8


class Results:
    """Static builders for well-formed :class:`CallToolResult` values."""

    def __init__(self) -> None:  # pragma: no cover - not meant to be instantiated
        raise TypeError("Results is a namespace of static helpers, not a type")

    @staticmethod
    def json(data: dict) -> CallToolResult:
        """Success result carrying pretty-printed JSON."""
        return CallToolResult(
            content=[TextContent(type="text", text=_json.dumps(data, indent=2))]
        )

    @staticmethod
    def text(message: str) -> CallToolResult:
        """Success result carrying plain text."""
        return CallToolResult(content=[TextContent(type="text", text=message)])

    @staticmethod
    def error(message: str) -> CallToolResult:
        """An error result (``is_error=True``) so the caller/agent sees the tool failed."""
        return CallToolResult(
            content=[TextContent(type="text", text=message)], is_error=True
        )

    @staticmethod
    def image(data: bytes, *, mime_type: str = "image/png") -> CallToolResult:
        """Image result, or an :meth:`error` when the bytes are missing/too small."""
        return Results._binary(data, mime_type, kind="image")

    @staticmethod
    def audio(data: bytes, *, mime_type: str = "audio/wav") -> CallToolResult:
        """Audio result (iCE speaks), or an :meth:`error` when empty/too small.

        Proves the piping is not PNG-only.
        """
        return Results._binary(data, mime_type, kind="audio")

    @staticmethod
    def bytes(data: bytes, *, mime_type: str, kind: str = "binary") -> CallToolResult:
        """Generic binary result carried as a base64 blob, guarded by :data:`MIN_BYTES`.

        Emitted as an :class:`ImageContent` when the mime is ``image/*``, otherwise as an
        :class:`AudioContent` (the two binary content shapes MCP v2 offers). The
        ``<=8``-byte guard is identical to :meth:`image` / :meth:`audio`.
        """
        return Results._binary(data, mime_type, kind=kind)

    @staticmethod
    def _binary(data: bytes, mime_type: str, *, kind: str) -> CallToolResult:
        if len(data) <= MIN_BYTES:
            return Results.error(
                f"empty or truncated {kind} ({len(data)} bytes) — "
                "the capture produced no usable data"
            )
        encoded = base64.b64encode(bytes(data)).decode("ascii")
        # image/* rides ImageContent; everything else rides AudioContent (the two
        # binary content kinds MCP v2 defines). A caller who wants audio calls audio().
        if kind == "image" or mime_type.startswith("image/"):
            block = ImageContent(type="image", data=encoded, mime_type=mime_type)
        else:
            block = AudioContent(type="audio", data=encoded, mime_type=mime_type)
        return CallToolResult(content=[block])

    @staticmethod
    def truncated_text(body: str, *, max_chars: int = 20000) -> CallToolResult:
        """Truncate ``body`` to ``max_chars``, marking how many chars were dropped so an
        agent never mistakes a partial payload for the whole thing."""
        if len(body) <= max_chars:
            return Results.text(body)
        dropped = len(body) - max_chars
        return Results.text(f"{body[:max_chars]}\n…[truncated {dropped} chars]")

File: kits/typescript/src/results.ts
// Content/result helpers — the generic PIPING an app tool rides on. The screenshot/audio
// CAPTURE stays app-side; ironmcp owns how raw bytes become a well-formed, guarded MCP result.
// No helper echoes a caller-supplied value. Pure: no SDK import, so the core stays trivially
// testable and robust across @modelcontextprotocol/sdk versions.

export type TextBlock = { type: "text"; text: string };
export type ImageBlock = { type: "image"; data: string; mimeType: string };
export type AudioBlock = { type: "audio"; data: string; mimeType: string };
export type ContentBlock = TextBlock | ImageBlock | AudioBlock;
export type ToolResult = { content: ContentBlock[]; isError?: boolean };

/** Minimum bytes that count as real payload. A WSLg/X11 capture can exit 0 yet emit an empty
 *  (<=8-byte) file; treat that as a failure, not media (loqu8 invariant #8). */
export const MIN_BYTES = 8;

type Bytes = Uint8Array | ArrayLike<number>;

function toBuffer(bytes: Bytes): Buffer {
  return Buffer.from(bytes as Uint8Array);
}

/** Success result carrying pretty-printed JSON. */
function json(data: Record<string, unknown>): ToolResult {
  return { content: [{ type: "text", text: JSON.stringify(data, null, 2) }] };
}

/** Success result carrying plain text. */
function text(message: string): ToolResult {
  return { content: [{ type: "text", text: message }] };
}

/** An error result (`isError: true`) so the caller/agent sees the tool failed. */
function error(message: string): ToolResult {
  return { content: [{ type: "text", text: message }], isError: true };
}

function binary(bytes: Bytes, mimeType: string, kind: "image" | "audio"): ToolResult {
  const buf = toBuffer(bytes);
  if (buf.length <= MIN_BYTES) {
    return error(
      `empty or truncated ${kind} (${buf.length} bytes) — the capture produced no usable data`,
    );
  }
  const data = buf.toString("base64");
  return { content: [{ type: kind, data, mimeType } as ContentBlock] };
}

/** Image result, or an [error] when the bytes are missing/too small (invariant #8). */
function image(bytes: Bytes, mimeType = "image/png"): ToolResult {
  return binary(bytes, mimeType, "image");
}

/** Audio result (iCE speaks), or an [error] when empty/too small. Proves the piping is not
 *  PNG-only. */
function audio(bytes: Bytes, mimeType = "audio/wav"): ToolResult {
  return binary(bytes, mimeType, "audio");
}

/** Truncate [body] to [maxChars], appending a marker naming how many chars were dropped, so an
 *  agent never mistakes a partial payload for the whole thing. */
function truncatedText(body: string, maxChars = 20000): ToolResult {
  if (body.length <= maxChars) return text(body);
  const dropped = body.length - maxChars;
  return text(`${body.slice(0, maxChars)}\n…[truncated ${dropped} chars]`);
}

export const Results = { json, text, error, image, audio, truncatedText };
